File: Co-Simulation/Sumo/sumo-1.7.0/tools/traci/_lane.py
# -*- coding: utf-8 -*-
# Eclipse SUMO, Simulation of Urban MObility; see https://eclipse.org/sumo
# Copyright (C) 2011-2020 German Aerospace Center (DLR) and others.
# This program and the accompanying materials are made available under the
# terms of the Eclipse Public License 2.0 which is available at
# https://www.eclipse.org/legal/epl-2.0/
# This Source Code may also be made available under the following Secondary
# Licenses when the conditions for such availability set forth in the Eclipse
# Public License 2.0 are satisfied: GNU General Public License, version 2
# or later which is available at
# https://www.gnu.org/licenses/old-licenses/gpl-2.0-standalone.html
# SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-or-later

# @file    _lane.py
# @author  Michael Behrisch
# @author  Daniel Krajzewicz
# @author  Laura Bieker
# @author  Jakob Erdmann
# @date    2011-03-17

from __future__ import absolute_import
from .domain import Domain
from . import constants as tc


def _readLinks(result):
    result.read("!Bi")  # Type Compound, Length
    nbLinks = result.readInt()
    links = []
    for _ in range(nbLinks):
        result.read("!B")                           # Type String
        approachedLane = result.readString()
        result.read("!B")                           # Type String
        approachedInternal = result.readString()
        result.read("!B")                           # Type Byte
        hasPrio = bool(result.read("!B")[0])
        result.read("!B")                           # Type Byte
        isOpen = bool(result.read("!B")[0])
        result.read("!B")                           # Type Byte
        hasFoe = bool(result.read("!B")[0])
        result.read("!B")                           # Type String
        state = result.readString()
        result.read("!B")                           # Type String
        direction = result.readString()
        result.read("!B")                           # Type Float
        length = result.readDouble()
        links.append((approachedLane, hasPrio, isOpen, hasFoe,
                      approachedInternal, state, direction, length))
    return links


_RETURN_VALUE_FUNC = {tc.LANE_LINKS: _readLinks}


class LaneDomain(Domain):

    def __init__(self):
        Domain.__init__(self, "lane", tc.CMD_GET_LANE_VARIABLE, tc.CMD_SET_LANE_VARIABLE,
                        tc.CMD_SUBSCRIBE_LANE_VARIABLE, tc.RESPONSE_SUBSCRIBE_LANE_VARIABLE,
                        tc.CMD_SUBSCRIBE_LANE_CONTEXT, tc.RESPONSE_SUBSCRIBE_LANE_CONTEXT,
                        _RETURN_VALUE_FUNC, subscriptionDefault=(tc.LAST_STEP_VEHICLE_NUMBER,))

    def getLength(self, laneID):
        """getLength(string) -> double

        Returns the length in m.
        """
        return self._getUniversal(tc.VAR_LENGTH, laneID)

    def getMaxSpeed(self, laneID):
        """getMaxSpeed(string) -> double

        Returns the maximum allowed speed on the lane in m/s.
        """
        return self._getUniversal(tc.VAR_MAXSPEED, laneID)

    def getWidth(self, laneID):
        """getWidth(string) -> double

        Returns the width of the lane in m.
        """
        return self._getUniversal(tc.VAR_WIDTH, laneID)

    def getAllowed(self, laneID):
        """getAllowed(string) -> list(string)

        Returns a list of allowed vehicle classes. An empty list means all vehicles are allowed.
        """
        return self._getUniversal(tc.LANE_ALLOWED, laneID)

    def getDisallowed(self, laneID):
        """getDisallowed(string) -> list(string)

        Returns a list of disallowed vehicle classes.
        """
        return self._getUniversal(tc.LANE_DISALLOWED, laneID)

    def getLinkNumber(self, laneID):
        """getLinkNumber(string) -> integer

        Returns the number of connections to successive lanes.
        """
        return self._getUniversal(tc.LANE_LINK_NUMBER, laneID)

    def getLinks(self, laneID, extended=True):
        """getLinks(string) -> list((string, bool, bool, bool))
        A list containing id of successor lane together with priority, open and foe
        for each link.
        if extended=True, each result tuple contains
        (string approachedLane, bool hasPrio, bool isOpen, bool hasFoe,
        string approachedInternal, string state, string direction, float length)
        """
        complete_data = self._getUniversal(tc.LANE_LINKS, laneID)
        if extended:
            return complete_data
        else:
            # for downward compatibility
            return [tuple(d[:4]) for d in complete_data]

    def getShape(self, laneID):
        """getShape(string) -> list((double, double))

        List of 2D positions (cartesian) describing the geometry.
        """
        return self._getUniversal(tc.VAR_SHAPE, laneID)

    def getEdgeID(self, laneID):
        """getEdgeID(string) -> string

        Returns the id of the edge the lane belongs to.
        """
        return self._getUniversal(tc.LANE_EDGE_ID, laneID)

    def getCO2Emission(self, laneID):
        """getCO2Emission(string) -> double

        Returns the CO2 emission in mg for the last time step on the given lane.
        """
        return self._getUniversal(tc.VAR_CO2EMISSION, laneID)

    def getCOEmission(self, laneID):
        """getCOEmission(string) -> double

        Returns the CO emission in mg for the last time step on the given lane.
        """
        return self._getUniversal(tc.VAR_COEMISSION, laneID)

    def getHCEmission(self, laneID):
        """getHCEmission(string) -> double

        Returns the HC emission in mg for the last time step on the given lane.
        """
        return self._getUniversal(tc.VAR_HCEMISSION, laneID)

    def getPMxEmission(self, laneID):
        """getPMxEmission(string) -> double

        Returns the particular matter emission in mg for the last time step on the given lane.
        """
        return self._getUniversal(tc.VAR_PMXEMISSION, laneID)

    def getNOxEmission(self, laneID):
        """getNOxEmission(string) -> double

        Returns the NOx emission in mg for the last time step on the given lane.
        """
        return self._getUniversal(tc.VAR_NOXEMISSION, laneID)

    def getFuelConsumption(self, laneID):
        """getFuelConsumption(string) -> double

        Returns the fuel consumption in ml for the last time step on the given lane.
        """
        return self._getUniversal(tc.VAR_FUELCONSUMPTION, laneID)

    def getNoiseEmission(self, laneID):
        """getNoiseEmission(string) -> double

        Returns the noise emission in db for the last time step on the given lane.
        """
        return self._getUniversal(tc.VAR_NOISEEMISSION, laneID)

    def getElectricityConsumption(self, laneID):
        """getElectricityConsumption(string) -> double

        Returns the electricity consumption in ml for the last time step.
        """
        return self._getUniversal(tc.VAR_ELECTRICITYCONSUMPTION, laneID)

    def getLastStepMeanSpeed(self, laneID):
        """getLastStepMeanSpeed(string) -> double

        Returns the average speed in m/s for the last time step on the given lane.
        """
        return self._getUniversal(tc.LAST_STEP_MEAN_SPEED, laneID)

    def getLastStepOccupancy(self, laneID):
        """getLastStepOccupancy(string) -> double

        Returns the occupancy in % for the last time step on the given lane.
        """
        return self._getUniversal(tc.LAST_STEP_OCCUPANCY, laneID)

    def getLastStepLength(self, laneID):
        """getLastStepLength(string) -> double

        Returns the mean vehicle length in m for the last time step on the given lane.
        """
        return self._getUniversal(tc.LAST_STEP_LENGTH, laneID)

    def getWaitingTime(self, laneID):
        """getWaitingTime() -> double

        .
        """
        return self._getUniversal(tc.VAR_WAITING_TIME, laneID)

    def getTraveltime(self, laneID):
        """getTraveltime(string) -> double

        Returns the estimated travel time in s for the last time step on the given lane.
        """
        return self._getUniversal(tc.VAR_CURRENT_TRAVELTIME, laneID)

    def getLastStepVehicleNumber(self, laneID):
        """getLastStepVehicleNumber(string) -> integer

        Returns the total number of vehicles for the last time step on the given lane.
        """
        return self._getUniversal(tc.LAST_STEP_VEHICLE_NUMBER, laneID)

    def getLastStepHaltingNumber(self, laneID):
        """getLastStepHaltingNumber(string) -> integer

        Returns the total number of halting vehicles for the last time step on the given lane.
        A speed of less than 0.1 m/s is considered a halt.
        """
        return self._getUniversal(tc.LAST_STEP_VEHICLE_HALTING_NUMBER, laneID)

    def getLastStepVehicleIDs(self, laneID):
        """getLastStepVehicleIDs(string) -> list(string)

        Returns the ids of the vehicles for the last time step on the given lane.
        """
        return self._getUniversal(tc.LAST_STEP_VEHICLE_ID_LIST, laneID)

    def getFoes(self, laneID, toLaneID):
        """getFoes(string, string) -> list(string)
        Returns the ids of incoming lanes that have right of way over the connection from laneID to toLaneID
        """
        return self._getUniversal(tc.VAR_FOES, laneID, "s", toLaneID)

    def getInternalFoes(self, laneID):
        """getFoes(string) -> list(string)
        Returns the ids of internal lanes that are in conflict with the given internal lane id
        """
        return self.getFoes(laneID, "")

    def setAllowed(self, laneID, allowedClasses):
        """setAllowed(string, list) -> None

        Sets a list of allowed vehicle classes. Setting an empty list means all vehicles are allowed.
        """
        if isinstance(allowedClasses, str):
            allowedClasses = [allowedClasses]
        self._setCmd(tc.LANE_ALLOWED, laneID, "l", allowedClasses)

    def setDisallowed(self, laneID, disallowedClasses):
        """setDisallowed(string, list) -> None

        Sets a list of disallowed vehicle classes.
        """
        if isinstance(disallowedClasses, str):
            disallowedClasses = [disallowedClasses]
        self._setCmd(tc.LANE_DISALLOWED, laneID, "l", disallowedClasses)

    def setMaxSpeed(self, laneID, speed):
        """setMaxSpeed(string, double) -> None

        Sets a new maximum allowed speed on the lane in m/s.
        """
        self._setCmd(tc.VAR_MAXSPEED, laneID, "d", speed)

    def setLength(self, laneID, length):
        """setLength(string, double) -> None

        Sets the length of the lane in m.
        """
        self._setCmd(tc.VAR_LENGTH, laneID, "d", length)

File: Co-Simulation/Sumo/sumo-1.7.0/tools/contributed/traas/src/main/java/de/tudresden/sumo/cmd/Lane.java
/****************************************************************************/
// Eclipse SUMO, Simulation of Urban MObility; see https://eclipse.org/sumo
// Copyright (C) 2017-2020 German Aerospace Center (DLR) and others.
// TraaS module
// Copyright (C) 2016-2017 Dresden University of Technology
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// https://www.eclipse.org/legal/epl-2.0/
// This Source Code may also be made available under the following Secondary
// Licenses when the conditions for such availability set forth in the Eclipse
// Public License 2.0 are satisfied: GNU General Public License, version 2
// or later which is available at
// https://www.gnu.org/licenses/old-licenses/gpl-2.0-standalone.html
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-or-later
/****************************************************************************/
/// @file    Lane.java
/// @author  Mario Krumnow
/// @author  Evamarie Wiessner
/// @date    2016
///
//
/****************************************************************************/
package de.tudresden.sumo.cmd;
import de.tudresden.sumo.config.Constants;
import de.tudresden.sumo.util.SumoCommand;
import de.tudresden.ws.container.SumoStringList;

/**
 *
 * @author Mario Krumnow
 * @author Evamarie Wiessner
 *
 */

public class Lane {

    //getter methods

    /**
     * Returns a list of allowed vehicle classes.
     *
     * @param laneID
     *            a string identifying a lane
     * @return list of allowed vehicle classes. An empty list means all vehicles
     *         are allowed.
     */

    public static SumoCommand getAllowed(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.LANE_ALLOWED, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_STRINGLIST);
    }

    /**
     * Returns the number of all lanes in the network.
     * @return the number of lanes
     */

    public static SumoCommand getIDCount() {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.ID_COUNT, "", Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_INTEGER);
    }

    /**
     * Returns the CO2 emissions (in mg) on the given lane during the last time
     * step.
     *
     * @param laneID
     *            a string identifying a lane
     * @return the CO2 emissions (in mg)
     */
    public static SumoCommand getCO2Emission(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_CO2EMISSION, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns the CO emissions (in mg) on the given lane during the last time
     * step.
     *
     * @param laneID
     *            a string identifying a lane
     * @return the CO emissions (in mg)
     */

    public static SumoCommand getCOEmission(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_COEMISSION, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns a list of disallowed vehicle classes.
     *
     * @param laneID
     *            a string identifying a lane
     * @return list of disallowed vehicle classes
     */

    public static SumoCommand getDisallowed(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.LANE_DISALLOWED, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_STRINGLIST);
    }

    /**
     * Returns the ID of the edge this lane belongs to.
     *
     * @param laneID
     *            a string identifying a lane
     * @return the id of the edge this lane belongs to
     */

    public static SumoCommand getEdgeID(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.LANE_EDGE_ID, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_STRING);
    }

    /**
     * Returns the electricity consumption in ml for the last time step.
     *
     * @param laneID
     *            a string identifying a lane
     *
     * @return the energy consumption
     */

    public static SumoCommand getElectricityConsumption(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_ELECTRICITYCONSUMPTION, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }


    /**
     * Returns the ids of incoming lanes that have right of way over the connection from laneID to toLaneID
     *
     * @param laneID
     *            a string identifying a lane
     * @param toLane
     *            a string identifying the target lane
     * @return the energy consumption
     */

    public static SumoCommand getFoes(String laneID, String toLane) {
        Object[] array = new Object[] {toLane};
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_FOES, laneID, array, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_STRINGLIST);
    }

    /**
     * Returns the ids of internal lanes that are in conflict with the given internal lane id
     *
     * @param laneID
     *            a string identifying a lane
     * @return the energy consumption
     */

    public static SumoCommand getInternalFoes(String laneID) {
        Object[] array = new Object[] {""};
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_FOES, laneID, array, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_STRINGLIST);
    }



    /**
     * Returns the fuel consumption (in ml) on the given lane during the last
     * time step.
     *
     * @param laneID
     *            a string identifying a lane
     * @return the fuel consumption (in ml)
     */

    public static SumoCommand getFuelConsumption(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_FUELCONSUMPTION, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns the HC emissions (in mg) on the given lane during the last time
     * step.
     *
     * @param laneID
     *            a string identifying a lane
     * @return the HC emissions (in mg)
     */

    public static SumoCommand getHCEmission(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_HCEMISSION, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns a list of IDs of all lanes in the network.
     *
     * @return a list of IDs of all lanes
     */

    public static SumoCommand getIDList() {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.TRACI_ID_LIST, "", Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_STRINGLIST);
    }

    /**
     * Returns the total number of halting vehicles on the given lane within the
     * last time step. A speed of less than 0.1 m/s is considered a halt.
     *
     * @param laneID
     *            a string identifying a lane
     * @return number of halting vehicles within the last time step
     */

    public static SumoCommand getLastStepHaltingNumber(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.LAST_STEP_VEHICLE_HALTING_NUMBER, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_INTEGER);
    }

    /**
     * Returns the total vehicle length (in m) on the given lane during the last
     * time step.
     *
     * @param laneID
     *            a string identifying a lane
     * @return vehicle length (in m) within the last time step
     */

    public static SumoCommand getLastStepLength(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.LAST_STEP_LENGTH, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns the mean speed (in m/s) on the given lane within the last time
     * step.
     *
     * @param laneID
     *            a string identifying a lane
     * @return the average speed (in m/s) within the last time step
     */

    public static SumoCommand getLastStepMeanSpeed(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.LAST_STEP_MEAN_SPEED, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns the occupancy (in %) on the given lane during the last time step.
     *
     * @param laneID
     *            a string identifying a lane
     * @return the percentage of time the lane was occupied
     */

    public static SumoCommand getLastStepOccupancy(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.LAST_STEP_OCCUPANCY, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns the IDs of the vehicles on the given lane within the last time
     * step.
     *
     * @param laneID
     *            a string identifying a lane
     * @return list of IDs of vehicles
     */

    public static SumoCommand getLastStepVehicleIDs(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.LAST_STEP_VEHICLE_ID_LIST, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_STRINGLIST);
    }

    /**
     * Returns number of vehicles on this lane within the last time step.
     *
     * @param laneID
     *            a string identifying a lane
     * @return number of vehicles
     */

    public static SumoCommand getLastStepVehicleNumber(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.LAST_STEP_VEHICLE_NUMBER, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_INTEGER);
    }

    /**
     * Returns the length (in m) of the named lane.
     *
     * @param laneID
     *            a string identifying a lane
     * @return length of the named lane (in m)
     */

    public static SumoCommand getLength(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_LENGTH, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns the number of links outgoing from this lane.
     *
     * @param laneID
     *            a string identifying a lane
     * @return number of connections to successive lanes
     */

    public static SumoCommand getLinkNumber(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.LANE_LINK_NUMBER, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_UBYTE);
    }

    /**
     * Returns descriptions of the links (in m) outgoing from this lane.
     *
     * @param laneID
     *            a string identifying a lane
     * @return a list containing IDs of sucessor lanes
     */

    public static SumoCommand getLinks(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.LANE_LINKS, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_COMPOUND);
    }

    /**
     * Returns the maximum speed (in m/s) allowed on this lane.
     *
     * @param laneID
     *            a string identifying a lane
     * @return maximum allowed speed (in m/s)
     */

    public static SumoCommand getMaxSpeed(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_MAXSPEED, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns the NOx emissions (in mg) on the given lane during the last time
     * step.
     *
     * @param laneID
     *            a string identifying a lane
     * @return the NOx emissions (in mg)
     */

    public static SumoCommand getNOxEmission(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_NOXEMISSION, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns the noise emissions (in db) on the given lane during the last
     * time step.
     *
     * @param laneID
     *            a string identifying a lane
     * @return the noise emissions (in mg)
     */

    public static SumoCommand getNoiseEmission(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_NOISEEMISSION, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns the chosen parameter
     *
     *  @param laneID a string identifying the edge
     *  @param param a string identifying the parameter
     *
     * @return the specific parameter
     */

    public static SumoCommand getParameter(String laneID, String param) {
        Object[] array = new Object[] {param};
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_PARAMETER, laneID, array, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_STRING);
    }

    /**
     * Sets the chosen parameter
     *
     *  @param vehID a string identifying the vehicle
     *  @param param a string identifying the parameter
     *  @param value a string identifying the new value
     *
     * @return SumoCommand
     */

    public static SumoCommand setParameter(String laneID, String param, String value) {
        Object[] array = new Object[] {param, value};
        return new SumoCommand(Constants.CMD_SET_LANE_VARIABLE, Constants.VAR_PARAMETER, laneID, array);
    }

    /**
     * Returns the particular matter emissions (in mg) on the given lane during
     * the last time step.
     *
     * @param laneID
     *            a string identifying a lane
     * @return the particular matter emissions (in mg)
     */

    public static SumoCommand getPMxEmission(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_PMXEMISSION, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns this lane's shape.
     *
     * @param laneID
     *            a string identifying a lane
     * @return list of position describing the geometry
     */

    public static SumoCommand getShape(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_SHAPE, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_POLYGON);
    }

    /**
     * Returns the estimated travel time (in s) on the given lane during the
     * last time step.
     *
     * @param laneID
     *            a string identifying a lane
     * @return estimated travel time (in s)
     */

    public static SumoCommand getTraveltime(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_CURRENT_TRAVELTIME, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns the width (in m) of the named lane.
     *
     * @param laneID
     *            a string identifying a lane
     * @return width of lane (in m)
     */

    public static SumoCommand getWidth(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_WIDTH, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * getWaitingTime
      * @param laneID
     *            a string identifying a lane
     * @return the waiting time of the lane (in s)
     */

    public static SumoCommand getWaitingTime(String laneID) {
        return new SumoCommand(Constants.CMD_GET_LANE_VARIABLE, Constants.VAR_WAITING_TIME, laneID, Constants.RESPONSE_GET_LANE_VARIABLE, Constants.TYPE_DOUBLE);
    }


    //setter methods

    /**
     * Set a list of allowed vehicle classes.
     *
     * @param laneID
     *            a string identifying a lane
     * @param allowedClasses
     *            allowed vehicle classes
     * @return SumoCommand
     */

    public static SumoCommand setAllowed(String laneID, SumoStringList allowedClasses) {

        return new SumoCommand(Constants.CMD_SET_LANE_VARIABLE, Constants.LANE_ALLOWED, laneID, allowedClasses);
    }

    /**
     * Set a list of disallowed vehicle classes.
     *
     * @param laneID
     *            a string identifying a lane
     * @param disallowedClasses
     *            not allowed vehicle classes
     * @return SumoCommand
     */

    public static SumoCommand setDisallowed(String laneID, SumoStringList disallowedClasses) {

        return new SumoCommand(Constants.CMD_SET_LANE_VARIABLE, Constants.LANE_DISALLOWED, laneID, disallowedClasses);
    }

    /**
     * Set the length (in m) of the named lane.
     *
     * @param laneID
     *            a string identifying a lane
     * @param length
     *            new length (in m) of the lane
     * @return SumoCommand
     */

    public static SumoCommand setLength(String laneID, double length) {

        return new SumoCommand(Constants.CMD_SET_LANE_VARIABLE, Constants.VAR_LENGTH, laneID, length);
    }

    /**
     * Set the maximum speed (in m/s) allowed on this lane.
     *
     * @param laneID
     *            a string identifying a lane
     * @param speed
     *            maximum velocity allowed (in m/s)
     * @return SumoCommand
     */

    public static SumoCommand setMaxSpeed(String laneID, double speed) {

        return new SumoCommand(Constants.CMD_SET_LANE_VARIABLE, Constants.VAR_MAXSPEED, laneID, speed);
    }


}
